design/ceph/nvmeof-recoverer.md

```diff
@@ -0,0 +1,184 @@
+---
+title: NVMe-oF OSD Recovery and Scheduling
+target-version: release-X.X
+---
+
+# NVMe-oF OSD Recovery and Scheduling
+
+## Summary
+We propose a recovery design, called NVMe-oF recoverer. NVMe-oF is an emerging storage network where the storage devices can be attached to any storage node using RDMA, TCP or Fiber channel. In this environment, the failure caused by a server node can be recoverable because the OSD can be re-spawned to another server node with the storage device attached to the failed server node through NVMe-oF.
+
+So, what we propose is a functionality that moves the OSD pod to another server node either automatically or manually when the server node fails because storage devices are still workable in NVMe-oF environment.
```

Comment:
This feature of moving OSDs dynamically between nodes is actually already a possibility with a [PVC-based cluster](https://rook.io/docs/rook/latest-release/CRDs/Cluster/pvc-cluster/). 
- This feature is commonly used in cloud environments where PVs can easily move between nodes. 
- The OSDs can be marked in the `storageClassDeviceSets` as `portable: true`. This allows the OSDs to move between nodes depending only on the restrictions of the underlying PV.
- When the OSDs are marked portable, the CRUSH map contains the PV name instead of the host name. Thus, when the OSDs move between hosts, the CRUSH map does not need to change. 

So this feature request may be a question more about setting up the PVs so Rook can just request them from a storage class, and perhaps no changes needed in Rook. Please consider that existing design to see if it already meets the requirements.

Reply:
Thanks for the detailed insights. My initial proposal was to achieve flexible, on-premise storage mobility and scalability by utilizing NVMe-oF directly, rather than relying solely on PVC-based provisioning. However, after revisiting your comments and examining the PVC-based portable OSD model, I've realized that it already provides a highly refined and stable management framework. This suggests that developing a dedicated NVMe-oF CSI driver might be more appropriate than integrating this functionality directly into Rook.
Proposed Enhancement
Current Issue: In the current PVC-based cluster design, the CRUSH map places OSDs under the root level based on their PV names. This can cause multiple PVs from the same physical storage server to appear as if they're in separate failure domains. As a result, Ceph may create replica sets entirely from OSDs supplied by a single storage server. If that server fails, the entire replica set could become unrecoverable.
Suggested Improvement: To address this, I propose adding a layer of topology awareness to the CRUSH map. Instead of relying solely on PV names, we could group PVs by their underlying storage servers. This would give Ceph a more accurate view of fault domains, improving fault tolerance and reducing the risk of catastrophic replica set failures in on-premise deployments.
What are your thoughts on this proposal? I'd appreciate any feedback on whether introducing a topology-aware hierarchy to the CRUSH map could strengthen the current design.

Reply:
The OSD topology can be controlled with the [topology labels](https://rook.io/docs/rook/latest-release/CRDs/Cluster/ceph-cluster-crd/?h=topology#osd-topology) on the nodes. Whenever the OSDs are created, the CRUSH map will include the zone/rack or other topology labels listed in that topic. The `host` topology is the only one that is affected by the pvc name. 

Typically you would have three zones with nodes belonging to each one, then OSDs would retain affinity to those zones. If the underlying PVs do not have the zone awareness, then you could create a separate `storageClassDeviceSet` with node affinity to a particular zone. This way the OSDs would belong to a specific zone. I believe this meets your request.

Reply:
In disaggregated architectures, where storage and compute nodes are separated, Rook currently exists a key challenge:
**Issue:** Although `portable mode` eliminates dependency on specific nodes, the CRUSH Map still relies on **node labels where the OSDs are running.** In an disaggregation architecture, multiple storage devices may be attached within the same storage node, which introduces a **new failure domain** distinct from the traditional server node-based failure domain. 
As a result, If the CRUSH Map is unaware of storage node failure domains, a storage node failure can simultaneously disconnect all attached devices, causing the storage system **to lose availability despite the replication strategy.**

![image](https://github.com/user-attachments/assets/e0f17494-28df-404c-ba60-f7fff029a0c3)
**Example Scenario (AS-IS):**
- `StorageNode0` and `StorageNode1` supply storage capacity via a dedicated CSI (e.g., `spdk-csi` or `mayastor`).
- OSDs run on `ComputeNode0` and `ComputeNode1` using PVC-based portable mode with **topology labels.**
- In this setup, the **CRUSH Map** relies on the node labels where the OSDs are deployed.
- The CRUSH Map distributes replicas across failure domains (e.g., OSD 3, 4 and 5).
- If `StorageNode1` crashes, only `OSD 4 and 5` fail, while OSD 2 on StorageNode0 remains accessible, preserving data availability.

**What’s Need (TO-BE):**
- Extend PVC-based portable mode so that the CRUSH Map uses the actual storage node topology (e.g., `StorageNode0` and `StorageNode1`) instead of compute node labels.
- Aligning the CRUSH Map with storage node topology ensures that replicas are evenly distributed across distinct physical storage nodes (e.g., `OSD 2, 4, 5`).
- If `StorageNode1` crashes, only the replicas on that node (OSD 4, 5) fail, while the remaining replica (OSD 2) on StorageNode0 remains accessible, ensuring data availability and improving fault tolerance.

By implementing this, CRUSH can ensure that replica sets are not placed on the same storage node, while the Kubernetes scheduler can optimize Pod placement based on computing resources. This enables a clear separation of storage and compute resources, enhancing fault tolerance and resource efficiency in disaggregated architectures.

Reply:
When OSDs are portable, a higher level of topology is required. For example, the nodes would be segregated into three zones. 
- OSDs are not portable across the zones. This ensures the data is safe across the zones. 
- OSDs are only portable across nodes within the same zone. 

This way, the data in the CRUSH map stays safe and available as long as one zone is down at a time.

Reply:
The [OSD topology](https://rook.io/docs/rook/latest-release/CRDs/Cluster/ceph-cluster-crd/#osd-topology) is based on node labels. In your diagram, I believe Rook is only aware of the node labels in the compute nodes, but not the storage nodes. Since you are mixing the connections between compute nodes and storage nodes, rook cannot separate the topology properly for the crush map. Rook would require you to have corresponding nodes between compute and storage nodes to maintain the topology separation. The topologies across those layers cannot be mixed. Perhaps you can try to configure the PV affinity to get the behavior you want, but keeping them separate is recommended to avoid the issues you raise.

Reply:
Thanks for your feedback. In our setup, Rook currently forces a 1:1 mapping between storage and compute nodes to properly reflect failure domains in the CRUSH map, which can be restrictive. I'd like to propose a new PR that makes `Rook` more flexible for PVC portable mode, ensuring storage nodes are accurately represented in the CRUSH map without requiring a strict 1:1 mapping.
Let me know what you think, and I'll move forward with the PR for further discussion.